streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -43,161 +42,106 @@
     }
 
-    private class KStreamKStreamJoinProcessor extends ContextualProcessor<K, V1, K, VOut> {
-        private WindowStore<K, V2> otherWindowStore;
-        private Sensor droppedRecordsSensor;
-        private Optional<KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V1, V2>>> outerJoinStore = Optional.empty();
+    protected abstract class KStreamKStreamJoinProcessor extends ContextualProcessor<K, VThis, K, VOut> {
         private InternalProcessorContext<K, VOut> internalProcessorContext;
+        private Sensor droppedRecordsSensor;
+
         private TimeTracker sharedTimeTracker;
+        private Optional<KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<VL, VR>>> outerJoinStore =
+                Optional.empty();
+        private WindowStore<K, VOther> otherWindowStore;
 
         @Override
         public void init(final ProcessorContext<K, VOut> context) {
             super.init(context);
             internalProcessorContext = (InternalProcessorContext<K, VOut>) context;
-
             final StreamsMetricsImpl metrics = (StreamsMetricsImpl) context.metrics();
-            droppedRecordsSensor = droppedRecordsSensor(Thread.currentThread().getName(), context.taskId().toString(), metrics);
+            droppedRecordsSensor =
+                    droppedRecordsSensor(Thread.currentThread().getName(), context.taskId().toString(), metrics);
             otherWindowStore = context.getStateStore(otherWindowName);
             sharedTimeTracker = sharedTimeTrackerSupplier.get(context.taskId());
 
             if (enableSpuriousResultFix) {
                 outerJoinStore = outerJoinWindowName.map(context::getStateStore);
 
                 sharedTimeTracker.setEmitInterval(
-                    StreamsConfig.InternalConfig.getLong(
-                        context.appConfigs(),
-                        EMIT_INTERVAL_MS_KSTREAMS_OUTER_JOIN_SPURIOUS_RESULTS_FIX,
-                        1000L
-                    )
+                        StreamsConfig.InternalConfig.getLong(
+                                context.appConfigs(),
+                                EMIT_INTERVAL_MS_KSTREAMS_OUTER_JOIN_SPURIOUS_RESULTS_FIX,
+                                1000L
+                        )
                 );
             }
         }
 
-        @SuppressWarnings("unchecked")
         @Override
-        public void process(final Record<K, V1> record) {
-
-            final long inputRecordTimestamp = record.timestamp();
-            final long timeFrom = Math.max(0L, inputRecordTimestamp - joinBeforeMs);
-            final long timeTo = Math.max(0L, inputRecordTimestamp + joinAfterMs);
-
-            sharedTimeTracker.advanceStreamTime(inputRecordTimestamp);
-
-            if (outer && record.key() == null && record.value() != null) {
-                context().forward(record.withValue(joiner.apply(record.key(), record.value(), null)));
+        public void process(final Record<K, VThis> thisRecord) {
```

Comment:
nit: Personally I don't think calling it thisRecord and thisRecordTimestamp are so necessary, because the type checker will make sure we're using the right one.